EVG-14894 mark tasks as generated while saving builds/tasks

ablack12 opened the pull request:
[EVG-14894](https://jira.mongodb.org/browse/EVG-14894)

### Description 
This race is probably best seen [here](https://mongodb.splunkcloud.com/en-US/app/search/search?q=search%20index%3Devergreen%20generate.tasks%20mongodb_mongo_v5.0_ubuntu1804_debug_ubsan_change_streams_sharded_collections_multiversion_passthrough_gen_bd09da168ea707379ffed826022dd14a48ca62f0_21_06_25_06_00_37%20%7C%20sort%20-metadata.time&display.page.search.mode=verbose&dispatch.sample_ratio=1&earliest=1624593600&latest=1624680000&workload_pool=standard_perf&sid=1626978073.2623590). What happens is job1 updates the parser project, and then it begins saving builds/tasks. While that's happening, job2 tries to update the parser project (because GeneratedTasks isn't set yet), but throws a validation error because the parser project _has_ already been updated. Job1 then finishes successfully and marks GeneratedTasks as true, but the task has already been failed by Job2.

I'm not sure if this is the best way to go about this, but I think that race errors typically come from UpdateParserProject. Once that's updated, I think it does make sense to consider the tasks "generated." An alternative is we could make a second boolean ProjectUpdated, but then we have to maintain more.

bsamek commented on model/generate.go:
```diff
@@ -161,8 +161,17 @@ func (g *GeneratedProject) Save(ctx context.Context, p *Project, pp *ParserProje
 
 	if err := g.saveNewBuildsAndTasks(ctx, v, p, t); err != nil {
+		// Because we were unable to save new builds and tasks, release the generator lock.
+		if err2 := task.MarkGeneratedTasks(t.Id, false); err2 != nil {
```
I think this trades one problem for another, but I don't have a good sense of which problem would be more common. I think the problem is that if there is an app crash or panic during `saveNewBuildsAndTasks`, then the task will be marked as having generated tasks despite not having generated. To be correct a lock like this would also need a lease, which is probably more complex than we want. (See https://martin.kleppmann.com/2016/02/08/how-to-do-distributed-locking.html for more detail about this.)

ablack12 replied:
Ah this is a good point. I think that problem would be less common, though still possible. I think the way to handle this instead then is to still retry on config errors, but we should retry less (either just for config errors or in general). I'll write a new thing.

ablack12 commented:
Updated description: This began to error when I introduced a change to not retry on validation error, because I mistakenly assumed this would never be a transient error, and it's confusing to users to see "Context canceled" instead of the actual problem. https://jira.mongodb.org/browse/EVG-14581

Since this is sometimes transient, it seems like we should log the error but still continue to retry in case it doesn't mean anything? I thought logging at Info level might be a good way to not make this seem like too serious of an error.

bsamek commented on agent/command/generate.go:
```diff
@@ -110,13 +110,13 @@ func (c *generateTask) Execute(ctx context.Context, comm client.Communicator, lo
 				fullErr := errors.New(strings.Join(generateStatus.Errors, ", "))
-				if strings.Contains(fullErr.Error(), evergreen.RetryGenerateTasksError) {
-					return true, fullErr
+				// if the error isn't related to saving the generated task, log it but still retry in case of race condition
```
What happens on the server side? I would have thought that a later job might also fail. Does this imply that a second, racing job will succeed?

ablack12 replied:
Ah I think it's not that a future racing job will succeed but that an in progress job that has only done some of the modification will succeed.

ablack12 replied:
That being said, I feel like we never get stuck in the situation of "this config has been modified but tasks/variants haven't been created" (i.e. if the app server crashed halfway through the job) which maybe implies we _could_ do the slightly bad lock from the previous commit. However that still has a chance of failing, whereas this might retry too many times in some cases but it won't fail incorrectly.

bsamek replied:
On the server side, does returning this error in `j.generate` cause the task to be marked as an error? What happens in that case?

ablack12 replied:
yes, although this was always true. We could erase the error from the task if the task is successfully generated in the future.
